samples/server.c

```diff
@@ -285,125 +286,120 @@ migration_device_state_transition(vfu_ctx_t *vfu_ctx, vfu_migr_state_t state)
+        
+        memcpy(buf, server_data->bar1 + read_start, length_in_bar1);
+        memcpy(buf + length_in_bar1, &server_data->bar0, length_in_bar0);
+    } else if (read_start >= server_data->bar1_size) {
+        // case 3: entire read lies within bar0
+        // TODO check the following is always allowed
+
+        read_start -= server_data->bar1_size;
+        read_end -= server_data->bar1_size;
+
+        memcpy(buf, &server_data->bar0 + read_start, bytes_read);
     }
-    server_data->migration.pending_bytes = 0;
 
-    return size;
+    server_data->migration.pending_bytes -= bytes_read;
+
+    return bytes_read;
 }
 
 static ssize_t
-migration_write_data(vfu_ctx_t *vfu_ctx, void *data,
-                     uint64_t size, uint64_t offset)
+migration_write_data(vfu_ctx_t *vfu_ctx, void *data, uint64_t size)
 {
     struct server_data *server_data = vfu_get_private(vfu_ctx);
     char *buf = data;
-    int ret;
 
     assert(server_data != NULL);
     assert(data != NULL);
 
-    if (offset != 0 || size < server_data->bar1_size) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "XXX bad migration data write %#llx-%#llx",
-                (unsigned long long)offset,
-                (unsigned long long)offset + size - 1);
-        errno = EINVAL;
-        return -1;
-    }
-
-    memcpy(server_data->bar1, buf, server_data->bar1_size);
-    buf += server_data->bar1_size;
-    size -= server_data->bar1_size;
-    if (size == 0) {
+    if (server_data->migration.pending_bytes == 0 || size == 0) {
         return 0;
     }
-    if (size != sizeof(server_data->bar0)) {
-        errno = EINVAL;
-        return -1;
-    }
-    memcpy(&server_data->bar0, buf, sizeof(server_data->bar0));
-    ret = bar0_access(vfu_ctx, buf, sizeof(server_data->bar0), 0, true);
-    assert(ret == (int)size); /* FIXME */
 
-    return 0;
-}
+    uint32_t total_write = server_data->bar1_size + sizeof(server_data->bar0);
 
+    uint32_t write_start = total_write - server_data->migration.pending_bytes;
+    uint32_t write_end = MIN(write_start + size, total_write); // exclusive
+    assert(write_end > write_start);
 
-static int
-migration_data_written(UNUSED vfu_ctx_t *vfu_ctx, UNUSED uint64_t count)
-{
-    /*
-     * We apply migration state directly in the migration_write_data callback,
-     * so we don't need to do anything here. We would have to apply migration
-     * state in this callback if the migration region was memory mappable, in
-     * which case we wouldn't know when the client wrote migration data.
-     */
+    uint32_t bytes_written = write_end - write_start;
 
-    return 0;
-}
+    if (write_end <= server_data->bar1_size) {
+        // case 1: entire write lies within bar1
+        // TODO check the following is always allowed
 
-static size_t
-nr_pages(size_t size)
-{
-    return (size / sysconf(_SC_PAGE_SIZE) +
-            (size % sysconf(_SC_PAGE_SIZE) > 1));
-}
+        memcpy(server_data->bar1 + write_start, buf, bytes_written);
+    } else if (
```

Comment:
fix formatting please?
If you find you need comments in the middle of an if, that's a good indication you need to think about a refactor!

Reply:
Yep, I'm not sure why I had the second case not just be an `else` at the end!